src/ixdat/techniques/deconvolution.py

```diff
@@ -220,30 +153,146 @@ def fs(s):
+                # print(kernel[i])
+            if norm:
+                area = np.trapz(kernel, t_kernel)
+                kernel = kernel / area
+            if matrix:
+                kernel = np.tile(kernel, (len(kernel), 1))
+                i = 1
+                while i < len(t_kernel):  # TODO: pythonize this?
+                    kernel[i] = np.concatenate((kernel[0][i:], kernel[0][:i]))
+                    i = i + 1
+        else:
+            raise TechniqueError(
+                "Cannot model impulse response if not initialized with parameters."
+            )
+        return t_kernel, kernel
 
-        elif self.type == "measured":
-            kernel = self.MS_data[1]
-            t_kernel = self.MS_data[0]
+    def calc_impulse_response_from_data(
+        self, dt=0.1, duration=100, tspan=None, tspan_bg=None, norm=True, matrix=False
+    ):
+        """Calculates impulse response from data.
+        TODO: add possibility of subtracting a linear background!!!
+        TODO: add option to plot the implulse response from data/ return an axis to
+            co-plot with the data
+        TODO: figure out if it's ok to just get rid of dt and duration (not used here)
+        TODO: change the name kernel to something else
 
-        if norm:
-            area = np.trapz(kernel, t_kernel)
-            kernel = kernel / area
+        Args:
+            tspan (list): tspan over which to calculate the impulse response
+            tspan_bg (list): tspan of background to subtract. if list of tspans is
+                passed, will interpolate between the points using
+                calc_linear_background()
+            norm (bool): If true the impulse response is normalized to its
+                area. Default is True.
+            matrix (bool): If true the circulant matrix constructed from the
+                impulse reponse is returned. Default is False.
+        """
+        if self.type == "measured":
+            if type(tspan_bg[0]) is not list:
+                t_kernel, kernel = self.data.grab_flux(
+                    mol=self.mol, tspan=tspan, tspan_bg=tspan_bg
+                )
+            else:
+                t_kernel, kernel_raw = self.data.grab_flux(mol=self.mol, tspan=tspan)
+                bg = calc_linear_background(t_kernel, kernel_raw, tspans=tspan_bg)
+                kernel = kernel_raw - bg
+            if norm:
+                area = np.trapz(kernel, t_kernel)
+                kernel = kernel / area
+            if matrix:
+                kernel = np.tile(kernel, (len(kernel), 1))
+                i = 1
+                while i < len(t_kernel):  # TODO: pythonize this?
+                    kernel[i] = np.concatenate((kernel[0][i:], kernel[0][:i]))
+                    i = i + 1
+        else:
+            raise TechniqueError("Cannot calculate impulse response without data.")
+        return t_kernel, kernel
+
+    @np.vectorize
+    def get_cond_number(sampling_freq, working_dist):
```

Comment:
A few things (I can see in the comment that the method is not finished):
- This method does not take `self` as the first argument. It should probably not be a method of the class, but a stand-alone function in the module.
- needs a lot more docstring!
- np.vectorize just implements a for-loop. There must be a better way. But I would have to understand what this method does to suggest one.
- What molecule is hard-coded and why?

Reply:
just added a todo link now. wont implement this